slacker/commands/emoji_list_command.py

```diff
@@ -0,0 +1,90 @@
+import os
+import json
+import requests
+
+from argparse import ArgumentParser
+
+from .command import Command
+from slacker.slack_api import SlackAPI
+
+class EmojiListCommand(Command):
+  def name(self):
+    return 'emoji.list'
+
+  def description(self):
+    return 'Lists custom emojis in workspace'
+
+  def aliases(self):
+    return ['emojis']
+
+  def make_parser(self):
+    parser = ArgumentParser(prog = self.name(), description = self.description())
+    parser.add_argument('-p', '--path', type = str,
+                        help='Local file path where to save emojis to')
+    parser.add_argument('-q', '--quiet', action='store_true',
+                        help="Don't print response from Slack API")
+    return parser
+
+  def print_emojis(self, emojis):
+    total = 0
+    for emoji in emojis:
+      print('{:<25} {:<50}'.format(emoji, emojis[emoji]))
+      total += 1
+
+    print('{} custom emojis'.format(total))
+
+  def __download_emojis(self, emojis):
+    save_data = {}
+
+    for emoji in emojis:
+      emoji_url = emojis[emoji]
+      if emoji_url[:5] == 'alias':
+        save_data[emoji] = emoji_url
+        continue
+
+      saved_file_path = self.__download_emoji(emoji, emojis[emoji])
+      save_data[emoji] = saved_file_path
+
+    self.__write_save_file(save_data)
+
+  def __download_emoji(self, emoji_name, url):
+    self.logger.debug('Downloading {}'.format(url))
+
+    res = requests.get(url, stream=True)
+    if res.status_code != 200:
+      self.logger.warning('Unable to download {}: {}'.format(emoji_name, url))
+
+    image_type = res.headers['Content-Type'].split('/')[1]
+    content_length = res.headers['Content-Length']
+
+    emoji_file = '{}.{}'.format(emoji_name, image_type)
+    local_save_path = os.path.join(self.local_save_path, emoji_file)
+
+    self.logger.debug('Writing to disk {}({}) -> {}'.format(url, content_length, self.local_save_path))
+    with open(local_save_path, 'wb') as f:
+      for chunk in res.iter_content(1024):
+        f.write(chunk)
+
+    return local_save_path
+
+  def __write_save_file(self, save_data):
+    with open(self.save_data_file, 'w') as f:
+      f.write(json.dumps(save_data, indent=2))
+
+    self.logger.info('Emoji image files saved to {}'.format(self.local_save_path))
+    cwd_save_data = os.path.join(os.getcwd(), self.save_data_file)
+    self.logger.info('Emoji save data file saved to {}'.format(cwd_save_data))
```

Comment:
Shouldn't the JSON file be saved to where `--path` points to if defined?

Reply:
I thought about it, but it is nice that it is saved relative to where you run slacker. The save data file also includes the path to the emoji was downloaded. Here what it looks like. Perhaps later we can create a command to do something with the emoji once they are downloaded, like import to another workspace or something, although the Slack API doesn't all this right now :/

```
{
  "bowtie": "/tmp/bowtie.png",
  "squirrel": "/tmp/squirrel.png",
  "glitch_crab": "/tmp/glitch_crab.png",
  "piggy": "/tmp/piggy.png",
  "cubimal_chick": "/tmp/cubimal_chick.png",
  "dusty_stick": "/tmp/dusty_stick.png",
  "slack": "/tmp/slack.png",
  "pride": "/tmp/pride.png",
  "thumbsup_all": "/tmp/thumbsup_all.gif",
  "slack_call": "/tmp/slack_call.png",
  "shipit": "alias:squirrel",
  "white_square": "alias:white_large_square",
  "black_square": "alias:black_large_square",
  "simple_smile": "/tmp/simple_smile.png",
  "metal": "/tmp/metal.png"
}
```

Reply:
That's a good point. Let's keep it as-is for now. It's a little annoying we can't import via their API, yes. :/ But I still think it's useful because it is then easier to add manually to another workspace.